cdm/core/src/test/java/ucar/nc2/util/cache/TestS3NetcdfFileCache.java

```diff
@@ -0,0 +1,147 @@
+
+  @BeforeClass
+  static public void setUp() {
+    cache = new FileCache(5, 100, 60 * 60);
+  }
+
+  @AfterClass
+  static public void tearDown() {
+    cache.clearCache(true);
+  }
+
+  static public class MyFileFactory implements FileFactory {
+    public FileCacheable open(DatasetUrl location, int buffer_size, CancelTask cancelTask, Object iospMessage)
+        throws IOException {
+      return NetcdfDatasets.openFile(location, buffer_size, cancelTask, iospMessage);
+    }
+  }
+
+  private int count = 0;
+
+  private void loadFileIntoCache(String s3url, FileCache cache) {
+    try {
+      DatasetUrl durl = DatasetUrl.create(null, s3url);
+      cache.acquire(factory, durl);
+      count++;
+    } catch (IOException e) {
+      fail(" *** failed on " + s3url);
+    }
+  }
+
+  @Test
+  @Category(NeedsUcarNetwork.class)
+  public void testS3NetcdfFileCache() throws IOException {
+    loadFileIntoCache(NCAR_G16_S3_URI, cache);
+    assertThat(cache.hits.get()).isEqualTo(0);
+    assertThat(cache.miss.get()).isEqualTo(1);
+    cache.showCache(new Formatter(System.out));
+
+    // count cache size
+    Map<Object, FileCache.CacheElement> map = cache.getCache();
+    assertThat(map.values().size()).isEqualTo(count);
+
+    for (Object key : map.keySet()) {
+      FileCache.CacheElement elem = map.get(key);
+      assertThat(elem.list.size()).isEqualTo(1);
+    }
+
+    // load same file again - should be added to the list, rather than creating a new elem
+    int saveCount = count;
+    loadFileIntoCache(NCAR_G16_S3_URI, cache);
+
+    map = cache.getCache();
+    cache.showCache(new Formatter(System.out));
+    // The file is currently locked, so no cache hit.
+    assertThat(cache.hits.get()).isEqualTo(0);
+    assertThat(cache.miss.get()).isEqualTo(2);
+    assertThat(map.values().size()).isEqualTo(saveCount); // problem
+
+    for (Object key : map.keySet()) {
+      FileCache.CacheElement elem = map.get(key);
+      assertThat(elem.list.size()).isEqualTo(2);
+      checkAllSame(elem.list);
+    }
+    cache.clearCache(true);
+
+    // Verify that cache hits actually happen.
+    // Load up the file and then close tthem (thus, unlocking it).
+    // Then load it up again.
+    // Since file is now unlocked, loading it up should result in a hit.
+    loadFileIntoCache(NCAR_G16_S3_URI, cache);
+    assertThat(cache.hits.get()).isEqualTo(0);
+    assertThat(cache.miss.get()).isEqualTo(3);
+    map = cache.getCache();
+    // close all
+    List<FileCacheable> files = new ArrayList<>();
+    for (Object key : map.keySet()) {
+      FileCache.CacheElement elem = map.get(key);
+      assertThat(elem.list.size()).isEqualTo(1);
+      for (FileCache.CacheElement.CacheFile file : elem.list) {
+        // Need to do it this way instead of directly closing them
+        // because closing the files changes the iterations
+        files.add(file.ncfile);
+      }
```

Comment:
Oh, I probably should add the same synchronization trick I had in the other test here too.